### PR TITLE
Advanced settings in reactxp

### PR DESCRIPTION
**Pull request by anderklander**
Also bumped reactxp version to allow multiple children to buttons

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/52)

### CODE REVIEW

**Comment by eriklarko** [comment]
*[app/components/AdvancedSettings.js, line 4 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4Kf92q-9M1FUfDtUNe:-L4Kf92q-9M1FUfDtUNf:b-k9csu0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cd3465ab91cb0bc40214f5b621265adcacf71fe2/app/components/AdvancedSettings.js#L4)):*
> ```JavaScript
> 
> import React from 'react';
> import { Component, Text, Button, View } from 'reactxp';
> ```

The reactxp buttons have a bad habit of changing the pointer to a cursor which we don't really want on the desktop. We created `app/components/styled/Button.js` to remove the pointer and provide a place to put default styles on buttons. We're using it in `app/components/Settings.js` if you want an example

```
...
import { Component, Text, View } from 'reactxp';
import { Button } from './styled';
....
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by eriklarko** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 100 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4KgO-hyAte3-KZDQVI:-L4KgO-iNv15tiTcIKBJ:b634kdk) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cd3465ab91cb0bc40214f5b621265adcacf71fe2/app/components/AdvancedSettings.js#L100)):*
> ```JavaScript
> 
>   _renderSelectedCell(value) {
>     return <Button style={[ styles.advanced_settings__cell, value === this.state.hoveredButtonIndex ? styles.advanced_settings__cell_selected_hover : null ]} onPress={ () => this.props.onSelect(value) } onHoverStart={() => this.handleButtonHover(value)} onHoverEnd={() => this.handleButtonHover(-1)} key={ value }>
> ```

We don't have any automatic formatting yet, but I find it easier to read these long components if there's one attribute per line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by eriklarko** [comment]
The height of the background is a little off.



---







*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-:-L4KjjM_iKlWZES1s9fT:b4bw29x)*

**Comment by eriklarko** [comment]
*[a discussion](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4KkZBQB8ML-CBtNNs2:-L4KkZBRvQFrC9LNmrYh:b-o3kc12) (no related file):*
I can't seem to scroll in the view

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by eriklarko** [comment]
Reviewed 1 of 2 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/components/AdvancedSettings.js, line 100 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4KgO-hyAte3-KZDQVI:-L4Kr4rDKovGfHjfEpaG:b-83sjoh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cd3465ab91cb0bc40214f5b621265adcacf71fe2/app/components/AdvancedSettings.js#L100)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

We don't have any automatic formatting yet, but I find it easier to read these long components if there's one attribute per line
</blockquote></details>

Nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by eriklarko** [comment]
Reviewed 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by anderklander** [comment]
Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/52#-L4KkZBQB8ML-CBtNNs2:-L4ZktOu5xofdmgwvIuY:b-rzyvo) (no related file):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

I can't seem to scroll in the view
</blockquote></details>

Fixed scrolling and the background.

---

*[app/components/AdvancedSettings.js, line 4 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/52#-L4Kf92q-9M1FUfDtUNe:-L4Zkqil2_rqWAG8nk0Y:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/cd3465ab91cb0bc40214f5b621265adcacf71fe2/app/components/AdvancedSettings.js#L4)):*
<details><summary><i>Previously, eriklarko (Erik Larkö) wrote…</i></summary><blockquote>

The reactxp buttons have a bad habit of changing the pointer to a cursor which we don't really want on the desktop. We created `app/components/styled/Button.js` to remove the pointer and provide a place to put default styles on buttons. We're using it in `app/components/Settings.js` if you want an example

```
...
import { Component, Text, View } from 'reactxp';
import { Button } from './styled';
....
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 74 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aHaeXgFTqzhoIOSK8:-L4aHaeXgFTqzhoIOSK9:bca39pg) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L74)):*
> ```JavaScript
>   constructor(props) {
>     super(props);
>     this.state = { hoveredButtonIndex: -1 };
> ```

Since constructor only sets initial state you could shorten this to:

```
class Selector extends Component {
  state = { ... }
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 63 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aHtkgxzjcUwEhQ529:-L4aHtkgxzjcUwEhQ52A:b-ub53cs) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L63)):*
> ```JavaScript
> }
> 
> class Selector extends React.Component {
> ```

Aren't we supposed to use `ReactXP.Component` instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 103 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aI_4mcQDGjMu6NAQq:-L4aI_4mcQDGjMu6NAQr:b-i8v5kw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L103)):*
> ```JavaScript
>     return <Button style={[ styles.advanced_settings__cell, value === this.state.hoveredButtonIndex ? styles.advanced_settings__cell_selected_hover : null ]}
>       onPress={ () => this.props.onSelect(value) }
>       onHoverStart={() => this.handleButtonHover(value)}
> ```

It feels that such rudimentary work should not be done in higher level components. Maybe we could add a prop that accepts a `hoverStyle` and merges them together with style then swaps them on hover automatically?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by anderklander** [comment]
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 63 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aHtkgxzjcUwEhQ529:-L4aKoA7eLJ1tUcVkx2t:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L63)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

Aren't we supposed to use `ReactXP.Component` instead?
</blockquote></details>

Done.

---

*[app/components/AdvancedSettings.js, line 74 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aHaeXgFTqzhoIOSK8:-L4aKncwy2tPSaP2G_Af:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L74)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

Since constructor only sets initial state you could shorten this to:

```
class Selector extends Component {
  state = { ... }
}
```
</blockquote></details>

Done.

---

*[app/components/AdvancedSettings.js, line 103 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aI_4mcQDGjMu6NAQq:-L4aKGiFm1LBdtjplQA2:b-pyl5v1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L103)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

It feels that such rudimentary work should not be done in higher level components. Maybe we could add a prop that accepts a `hoverStyle` and merges them together with style then swaps them on hover automatically?
</blockquote></details>

Make a component of the styled button we have and just move the functionality there?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
Reviewed 3 of 5 files at r1, 1 of 2 files at r2, 2 of 2 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 103 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aI_4mcQDGjMu6NAQq:-L4aN2UbEPYb7yWcxJ45:b-uuhhjc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L103)):*
<details><summary><i>Previously, anderklander wrote…</i></summary><blockquote>

Make a component of the styled button we have and just move the functionality there?
</blockquote></details>

Well think about it. We have a button in every view and a lot of them have hover style. Maybe it would make sense to incorporate a hover logic into it? Throwing an idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by anderklander** [comment]
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[app/components/AdvancedSettings.js, line 103 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-L4aI_4mcQDGjMu6NAQq:-L4aNVECGdcNh50CNcaN:bavjv9o) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d417f0e355cd95d4d0bbd3735cd6ca520554cbdc/app/components/AdvancedSettings.js#L103)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

Well think about it. We have a button in every view and a lot of them have hover style. Maybe it would make sense to incorporate a hover logic into it? Throwing an idea.
</blockquote></details>

I would like to have more components, there are just a few basic button types in the app and it feels like it would be nice to have them as components, could simplify the styling a bit as well ... Perhaps take it in a future PR? Should not take so long but i guess we could spend some time in a discussion on how we should implement/use them?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52)*

**Comment by pronebird** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-:-L4aPtYuDSzBFFD6EEjt:bnfp4nl)*

**Comment by eriklarko** [comment]
Very nice! :lgtm_strong: 

---

Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/52#-:-L4g1L3S_tbOKEbObj00:b-il1ii9)*
